File: platform/polyglot/cache_json_awss3_storage/mjs/src/client-factory-with-http-client.ts
/**
 * S3 Client Factory with HTTP Client Injection
 *
 * Companion to client-factory.ts that accepts an undici Dispatcher from
 * fetch_client as the transport layer instead of the default NodeHttpHandler.
 * This enables unified proxy/SSL/timeout configuration and connection pooling
 * across the codebase.
 *
 * Polyglot parity with Python client_factory_with_http_client.py.
 *
 * @example
 * ```typescript
 * import { Agent } from "undici";
 * import {
 *   getClientFactoryWithHttpClient,
 *   ClientAsyncWithHttpClient,
 *   createStorage,
 * } from "cache_json_awss3_storage";
 *
 * const dispatcher = new Agent({ connect: { timeout: 10_000 } });
 * const config = getClientFactoryWithHttpClient({
 *   bucketName: "my-bucket",
 *   regionName: "us-east-1",
 * });
 *
 * const handle = new ClientAsyncWithHttpClient(config, dispatcher);
 * try {
 *   const storage = createStorage({ s3Client: handle.client, bucketName: config.bucketName });
 *   await storage.save("key", { data: "value" });
 * } finally {
 *   handle.destroy();
 * }
 * ```
 */

import { S3Client } from "@aws-sdk/client-s3";
import type { Dispatcher } from "undici";
import type { ClientConfig, ClientFactoryOptions, AsyncClientHandle } from "./client-factory.js";
import { JsonS3StorageConfigError } from "./errors.js";
import { create as createLogger } from "./logger.js";
import type { S3ClientInterface } from "./types.js";
import { UndiciHttpHandler } from "./undici-http-handler.js";

const logger = createLogger("cache_json_awss3_storage", import.meta.url);

/**
 * Create an S3 ClientConfig from connection parameters for use with
 * an injected HTTP client.
 *
 * Same validation as getClientFactory. The returned config is used with
 * ClientAsyncWithHttpClient or ClientSyncWithHttpClient.
 *
 * Polyglot parity with Python get_client_factory_with_http_client().
 *
 * @param options - Connection parameters (same as getClientFactory)
 * @returns ClientConfig instance
 *
 * @example
 * ```typescript
 * const config = getClientFactoryWithHttpClient({
 *   bucketName: "my-cache-bucket",
 *   regionName: "us-east-1",
 * });
 *
 * const handle = new ClientAsyncWithHttpClient(config, dispatcher);
 * ```
 */
export function getClientFactoryWithHttpClient(
  options: ClientFactoryOptions
): ClientConfig {
  if (!options.bucketName) {
    throw new JsonS3StorageConfigError("bucketName is required");
  }

  const type = options.type ?? "s3";
  if (type !== "s3") {
    throw new JsonS3StorageConfigError(
      `Unsupported client type: "${type}", expected "s3"`
    );
  }

  const config: ClientConfig = {
    bucketName: options.bucketName,
    proxyUrl: options.proxyUrl,
    endpointUrl: options.endpointUrl,
    awsSecretAccessKey: options.awsSecretAccessKey,
    awsAccessKeyId: options.awsAccessKeyId,
    regionName: options.regionName,
    addressingStyle: options.addressingStyle ?? "path",
    connectionTimeout: options.connectionTimeout ?? 20,
    readTimeout: options.readTimeout ?? 60,
    retriesMaxAttempts: options.retriesMaxAttempts ?? 3,
    type,
    verify: options.verify ?? true,
    ttl: options.ttl ?? 600,
  };

  logger.info(
    `getClientFactoryWithHttpClient: created config bucket=${config.bucketName}, ` +
      `region=${config.regionName}, endpoint=${config.endpointUrl}, ttl=${config.ttl}`
  );

  return config;
}

/**
 * Build an S3Client using the UndiciHttpHandler as the transport layer.
 */
function buildS3ClientWithHttpClient(
  config: ClientConfig,
  dispatcher: Dispatcher
): S3Client {
  const clientConfig: Record<string, unknown> = {};

  if (config.regionName) {
    clientConfig.region = config.regionName;
  }

  if (config.endpointUrl) {
    clientConfig.endpoint = config.endpointUrl;
    clientConfig.forcePathStyle = config.addressingStyle === "path";
  }

  if (config.awsAccessKeyId && config.awsSecretAccessKey) {
    clientConfig.credentials = {
      accessKeyId: config.awsAccessKeyId,
      secretAccessKey: config.awsSecretAccessKey,
    };
  }

  if (config.verify === false) {
    clientConfig.tls = false;
  }

  clientConfig.requestHandler = new UndiciHttpHandler(dispatcher, {
    connectionTimeout: (config.connectionTimeout ?? 20) * 1000,
    requestTimeout: (config.readTimeout ?? 60) * 1000,
  });

  clientConfig.maxAttempts = config.retriesMaxAttempts ?? 3;

  return new S3Client(clientConfig as any);
}

/**
 * Async S3 client wrapper using an undici Dispatcher.
 *
 * Polyglot parity with Python ClientAsyncWithHttpClient(config, httpx_client).
 * The caller owns the Dispatcher lifecycle — destroy() only tears down the S3Client.
 *
 * @example
 * ```typescript
 * const handle = new ClientAsyncWithHttpClient(config, dispatcher);
 * try {
 *   const storage = createStorage({ s3Client: handle.client, bucketName: config.bucketName });
 *   await storage.save("key", { data: "value" });
 * } finally {
 *   handle.destroy();
 * }
 *
 * // Or with Symbol.asyncDispose (Node 20+ / TypeScript 5.2+):
 * await using handle = new ClientAsyncWithHttpClient(config, dispatcher);
 * ```
 */
export class ClientAsyncWithHttpClient {
  /** The S3Client instance, compatible with S3ClientInterface */
  readonly client: S3Client & S3ClientInterface;
  /** The config used to create the client */
  readonly config: ClientConfig;

  constructor(config: ClientConfig, dispatcher: Dispatcher) {
    if (config.type !== "s3") {
      throw new JsonS3StorageConfigError(
        `Unsupported client type: "${config.type}", expected "s3"`
      );
    }

    if (!dispatcher) {
      throw new JsonS3StorageConfigError("dispatcher is required");
    }

    this.config = config;
    this.client = buildS3ClientWithHttpClient(config, dispatcher) as S3Client &
      S3ClientInterface;

    logger.debug(
      `ClientAsyncWithHttpClient: opened connection ` +
        `endpoint=${config.endpointUrl}, ` +
        `region=${config.regionName}, ` +
        `bucket=${config.bucketName}`
    );
  }

  /** Destroy the S3Client and release resources. The Dispatcher is NOT closed. */
  destroy(): void {
    try {
      this.client.destroy();
      logger.debug("ClientAsyncWithHttpClient: closed connection");
    } catch {
      // ignore destroy errors
    }
  }

  /** Support for `await using handle = new ClientAsyncWithHttpClient(...)` */
  async [Symbol.asyncDispose](): Promise<void> {
    this.destroy();
  }
}

/**
 * Sync S3 client wrapper using an undici Dispatcher.
 *
 * In Node.js all S3 operations are async-native. This class is an alias
 * for ClientAsyncWithHttpClient, provided for polyglot API parity
 * with Python's ClientSyncWithHttpClient(config, httpx_client).
 *
 * @example
 * ```typescript
 * const handle = new ClientSyncWithHttpClient(config, dispatcher);
 * try {
 *   // still async under the hood — Node.js is async-native
 *   await handle.client.send(new GetObjectCommand({ ... }));
 * } finally {
 *   handle.destroy();
 * }
 * ```
 */
export class ClientSyncWithHttpClient extends ClientAsyncWithHttpClient {
  constructor(config: ClientConfig, dispatcher: Dispatcher) {
    super(config, dispatcher);
  }
}

/**
 * Create an S3Client from a ClientConfig using an undici Dispatcher.
 *
 * Convenience function wrapping ClientAsyncWithHttpClient.
 *
 * @param config - ClientConfig from getClientFactoryWithHttpClient
 * @param dispatcher - undici Dispatcher/Agent from fetch_client
 * @returns AsyncClientHandle with client, destroy, and config
 */
export function createAsyncClientWithHttpClient(
  config: ClientConfig,
  dispatcher: Dispatcher
): AsyncClientHandle {
  const handle = new ClientAsyncWithHttpClient(config, dispatcher);
  return {
    client: handle.client,
    destroy: () => handle.destroy(),
    config: handle.config,
  };
}

/**
 * Create an S3Client from a ClientConfig using an undici Dispatcher (sync alias).
 *
 * In Node.js, all S3 operations are async-native. This function is an alias
 * for createAsyncClientWithHttpClient, provided for polyglot API parity
 * with Python's ClientSyncWithHttpClient.
 *
 * @param config - ClientConfig from getClientFactoryWithHttpClient
 * @param dispatcher - undici Dispatcher/Agent from fetch_client
 * @returns AsyncClientHandle with client, destroy, and config
 */
export function createSyncClientWithHttpClient(
  config: ClientConfig,
  dispatcher: Dispatcher
): AsyncClientHandle {
  return createAsyncClientWithHttpClient(config, dispatcher);
}

File: platform/polyglot/cache_json_awss3_storage/py/cache_json_awss3_storage/client_factory_with_http_client.py
"""
S3 Client Factory with HTTP Client Injection

Companion to client_factory.py that accepts httpx AsyncClient/Client from
fetch_client as the transport layer instead of native aiobotocore/boto3.
This enables unified proxy, SSL, timeout, and connection pooling configuration
across the codebase.

Polyglot parity with TypeScript client-factory-with-http-client.ts.

Usage:
    import httpx
    from cache_json_awss3_storage import (
        get_client_factory_with_http_client,
        ClientAsyncWithHttpClient,
        ClientSyncWithHttpClient,
        create_storage,
    )

    config = get_client_factory_with_http_client(
        bucket_name="my-bucket",
        region_name="us-east-1",
    )

    # Async usage
    async with httpx.AsyncClient() as http:
        async with ClientAsyncWithHttpClient(config, http) as s3:
            storage = create_storage(s3, config.bucket_name, ttl=config.ttl)
            await storage.save("key", {"data": "value"})

    # Sync usage
    with httpx.Client() as http:
        with ClientSyncWithHttpClient(config, http) as s3:
            s3.put_object(Bucket=config.bucket_name, Key="k", Body=b"data")
"""

from __future__ import annotations

from typing import Any

import httpx

from cache_json_awss3_storage.exceptions import JsonS3StorageConfigError
from cache_json_awss3_storage.httpx_s3_client import (
    HttpxS3ClientAsync,
    HttpxS3ClientSync,
)
from cache_json_awss3_storage.logger import create as create_logger
from cache_json_awss3_storage.models import ClientConfig

_logger = create_logger("cache_json_awss3_storage", __file__)


class ClientAsyncWithHttpClient:
    """
    Async S3 client context manager using httpx.AsyncClient.

    Yields an HttpxS3ClientAsync compatible with S3ClientProtocol.
    The caller owns the httpx.AsyncClient lifecycle.

    Example:
        async with httpx.AsyncClient() as http:
            async with ClientAsyncWithHttpClient(config, http) as s3:
                await s3.put_object(Bucket="b", Key="k", Body=b"data")
    """

    def __init__(
        self, config: ClientConfig, httpx_client: httpx.AsyncClient
    ) -> None:
        if config.type != "s3":
            raise JsonS3StorageConfigError(
                f"Unsupported client type: {config.type!r}, expected 's3'"
            )
        self._config = config
        self._httpx_client = httpx_client
        self._client: HttpxS3ClientAsync | None = None

    async def __aenter__(self) -> HttpxS3ClientAsync:
        self._client = HttpxS3ClientAsync(self._config, self._httpx_client)

        _logger.debug(
            f"ClientAsyncWithHttpClient: opened connection "
            f"endpoint={self._config.endpoint_url}, "
            f"region={self._config.region_name}, "
            f"bucket={self._config.bucket_name}"
        )
        return self._client

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        # No-op: caller owns the httpx client lifecycle
        self._client = None
        _logger.debug("ClientAsyncWithHttpClient: released client reference")


class ClientSyncWithHttpClient:
    """
    Sync S3 client context manager using httpx.Client.

    Yields an HttpxS3ClientSync with sync methods for S3 operations.
    The caller owns the httpx.Client lifecycle.

    Example:
        with httpx.Client() as http:
            with ClientSyncWithHttpClient(config, http) as s3:
                s3.put_object(Bucket="b", Key="k", Body=b"data")
    """

    def __init__(
        self, config: ClientConfig, httpx_client: httpx.Client
    ) -> None:
        if config.type != "s3":
            raise JsonS3StorageConfigError(
                f"Unsupported client type: {config.type!r}, expected 's3'"
            )
        self._config = config
        self._httpx_client = httpx_client
        self._client: HttpxS3ClientSync | None = None

    def __enter__(self) -> HttpxS3ClientSync:
        self._client = HttpxS3ClientSync(self._config, self._httpx_client)

        _logger.debug(
            f"ClientSyncWithHttpClient: opened connection "
            f"endpoint={self._config.endpoint_url}, "
            f"region={self._config.region_name}, "
            f"bucket={self._config.bucket_name}"
        )
        return self._client

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        # No-op: caller owns the httpx client lifecycle
        self._client = None
        _logger.debug("ClientSyncWithHttpClient: released client reference")


def get_client_factory_with_http_client(
    *,
    bucket_name: str,
    proxy_url: str | None = None,
    endpoint_url: str | None = None,
    aws_secret_access_key: str | None = None,
    aws_access_key_id: str | None = None,
    region_name: str | None = None,
    addressing_style: str = "path",
    connection_timeout: int = 20,
    read_timeout: int = 60,
    retries_max_attempts: int = 3,
    type: str = "s3",
    verify: bool = True,
    ttl: float = 600.0,
) -> ClientConfig:
    """
    Create an S3 ClientConfig for use with httpx-based client context managers.

    Same validation as get_client_factory. The returned config is used with
    ClientAsyncWithHttpClient or ClientSyncWithHttpClient to obtain a
    ready-to-use S3 client backed by httpx transport.

    Args:
        bucket_name: Target S3 bucket name (required)
        proxy_url: HTTPS proxy URL for S3 connections
        endpoint_url: Custom S3-compatible endpoint (e.g. MinIO, LocalStack)
        aws_secret_access_key: AWS secret access key (falls back to env/profile)
        aws_access_key_id: AWS access key ID (falls back to env/profile)
        region_name: AWS region name
        addressing_style: S3 addressing style ("path" or "virtual")
        connection_timeout: Connection timeout in seconds
        read_timeout: Read timeout in seconds
        retries_max_attempts: Maximum retry attempts for transient failures
        type: Client type, must be "s3"
        verify: Enable SSL certificate verification
        ttl: Default TTL for cached entries in seconds

    Returns:
        ClientConfig instance for use with ClientAsyncWithHttpClient
        or ClientSyncWithHttpClient

    Example:
        config = get_client_factory_with_http_client(
            bucket_name="my-cache-bucket",
            region_name="us-east-1",
            endpoint_url="http://localhost:4566",
            ttl=3600.0,
        )

        async with httpx.AsyncClient() as http:
            async with ClientAsyncWithHttpClient(config, http) as s3:
                storage = create_storage(s3, config.bucket_name, ttl=config.ttl)
                await storage.save("key", {"data": "value"})
    """
    if not bucket_name:
        raise JsonS3StorageConfigError("bucket_name is required")

    if type != "s3":
        raise JsonS3StorageConfigError(
            f"Unsupported client type: {type!r}, expected 's3'"
        )

    config = ClientConfig(
        bucket_name=bucket_name,
        proxy_url=proxy_url,
        endpoint_url=endpoint_url,
        aws_secret_access_key=aws_secret_access_key,
        aws_access_key_id=aws_access_key_id,
        region_name=region_name,
        addressing_style=addressing_style,
        connection_timeout=connection_timeout,
        read_timeout=read_timeout,
        retries_max_attempts=retries_max_attempts,
        type=type,
        verify=verify,
        ttl=ttl,
    )

    _logger.info(
        f"get_client_factory_with_http_client: created config "
        f"bucket={bucket_name}, region={region_name}, "
        f"endpoint={endpoint_url}, ttl={ttl}"
    )

    return config
